src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs

```diff
@@ -125,22 +131,57 @@ public async Task ExecuteAsync(ExportJobRecord exportJobRecord, WeakETag weakETa
+                    // 1. During update of the job record after the changes have been committed.
+                    // 2. During periodic check of the job record.
+                    //
+                    // If the cancellation is detected during step 1, it means the Output in the persisted job record is out of sync since it failed to update the job record.
+                    // In that case, we would want to update the job record so that it matches (best effort) what's been actually committed.
+                    // If the cancellation is detected during step 2, we don't need to update the job record even though there are new resources
+                    // exported because those resources haven't been committed yet and therefore the user will not see them.
                     if (progress.Page % _exportJobConfiguration.NumberOfPagesPerCommit == 0)
                     {
+                        // Commit the changes.
                         await _exportDestinationClient.CommitAsync(cancellationToken);
 
                         // Update the job record.
-                        await UpdateAndCommitJobRecord(_exportJobRecord, cancellationToken);
+                        try
+                        {
+                            await UpdateJobRecordAsync(cancellationToken);
+                        }
+                        catch (JobConflictException)
+                        {
+                            await HandleExportJobUpdatedExternallyAsync(updatedExportJob: null, updateJobRecord: true, cancellationToken);
+
+                            return;
+                        }
 
                         currentBatchId = progress.Page;
                     }
+                    else
+                    {
+                        bool updated = false;
+                        ExportJobOutcome updatedExportJob = null;
+
+                        // Check to see if the job has been canceled.
+                        using (IScoped<IFhirOperationDataStore> fhirOperationDataStore = _fhirOperationDataStoreFactory())
+                        {
+                            (updated, updatedExportJob) = await fhirOperationDataStore.Value.TryGetUpdatedExportJobAsync(_exportJobRecord.Id, _weakETag, cancellationToken);
+                        }
+
+                        if (updated)
+                        {
+                            await HandleExportJobUpdatedExternallyAsync(updatedExportJob, updateJobRecord: false, cancellationToken);
+
+                            return;
+                        }
+                    }
                 }
 
                 // Commit one last time for any pending changes.
                 await _exportDestinationClient.CommitAsync(cancellationToken);
 
-                await UpdateAndCommitJobStatus(OperationStatus.Completed, updateEndTimestamp: true, cancellationToken);
+                await CompleteJobAsync(OperationStatus.Completed, cancellationToken);
```

Comment:
This could still throw a JobConflictException since we will be updating the JobRecord during this stage.

Reply:
Chat about this a little bit. Removed the logic that tries to keep the output up-to-date.